include/oneapi/dpl/pstl/ranges/nanorange.hpp

```diff
@@ -4068,6 +4068,130 @@ make_reverse_iterator(I i)
     return reverse_iterator<I>(::std::move(i));
 }
 
+template <typename I>
+class offset_iterator
+{
+  public:
+    using iterator_type = I;
+    using difference_type = iter_difference_t<I>;
+    using value_type = iter_value_t<I>;
+    using iterator_category = detail::legacy_iterator_category_t<I>;
+    using reference = iter_reference_t<I>;
+    using pointer = I;
+
+    offset_iterator() = default;
+    offset_iterator(const offset_iterator&) = default;
+    explicit offset_iterator(I it, difference_type n, difference_type offset, I b) : current_(it), beg_(b), n_(n), offset_(offset)
+    {
+        assert(it >= b);
+    }
+
+    reference operator*() const 
+    {
+        auto i = current_ - beg_;
+        return *(beg_ + (i + offset_)%n_); 
```

Comment:
I think about `operator*` implementation and if I understand everything correctly I tend to disagree with that implementation because seems like `current_` may point to somewhere that doesn't belong to this sequence (e.g. due to `operator+=`) and then everything just works because of remainder. Should we reconsider to put this logic into `operator++` and the similar ones?

Reply:
Actually, I thought about that... But yes, it always does work because of remainder :-)
Otherwise, if wee add the logic to `operator+=` we have got more complicated code causing we have to check bound on each `+` . Furthermore, we also will not be able to unambiguous calculate a difference between iterators.